### ISSUE TITLE
Diffing values

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nix#1164]
When doing my nixpkgs refactors, I'm often trying to avoid changing any hashes. To help me diagnose why a hash changes, it would be really nice to be able to compare two nix values and do a tree diff.

For example, comparing `{ a = { q = null; }; }` and `{ a = { }; b = [ 2 ]; }`, I'd love something like:
```nix
{
  a = {
    -q = null;
  };
  +b = [ 2 ];
}
```

It's possible a nixpkgs lib function would suffice, in which case I'll reopen the issue there.

**Comment by Ericson2314** [comment]
One thing that would be nice to toggle is whether to recur into derivations or just treat them as an atom. Ideally I'd be able to say "always atomic" "never atomic" or "*n* derivations deep non-atomic".

**Comment by vcunat** [comment]
Several times I diffed the *.drv files and "recursed manually" into suspicious differences. Some tool for this would be nice to have, though I suspect it would be tricky to cover more complex cases reasonably...

**Comment by regnat** [comment]
Wouldn't something using `toJSON` (or `toXML`, but this would probably be less readable) and then a json diffing tool be sufficient for most cases and simpler to implement than rewriting this in nix ?

Of course the call to `toJSON` would still need to be wrapped to prevent unwanted stuff from being evaluated − if you want to diff the entire nixpkgs repo for example −, but that should be feasible I think

**Comment by Ericson2314** [comment]
@regnat Good point. I was going to raise a concern about laziness and cycles, but that is a problem with my original proposal too. Pruning at a certain depth (or, even better, an attempt to break cycles along the lines of the infinite recursion check) would solve that.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
